validator/server/src/main/java/org/ethereum/beacon/validator/api/ValidatorRest.java

```diff
@@ -0,0 +1,389 @@
+              syncingResponse =
+                  new SyncingResponse(
+                      syncStatus.isSyncing(),
+                      syncStatus.getStart() == null
+                          ? BigInteger.ZERO
+                          : syncStatus.getStart().toBI(),
+                      syncStatus.getCurrent() == null
+                          ? BigInteger.ZERO
+                          : syncStatus.getCurrent().toBI(),
+                      syncStatus.getBestKnown() == null
+                          ? BigInteger.ZERO
+                          : syncStatus.getBestKnown().toBI());
+            });
+  }
+
+  @Override
+  HttpServer addRoutes(HttpServer httpServer) {
+    return httpServer.route(
+        routes ->
+            routes
+                .get("/node/version", doJsonResponse(this::produceVersionResponse))
+                .get("/node/genesis_time", doJsonResponse(this::produceGenesisTimeResponse))
+                .get("/node/syncing", doJsonResponse(this::produceSyncingResponse))
+                .get(
+                    "/validator/duties",
+                    processGetRequestImpl(this::produceValidatorDutiesResponse))
+                .get("/validator/block", processGetRequestImpl(this::produceValidatorBlockResponse))
+                .post("/validator/block", processPostRequestImpl(this::acceptBlockSubmit))
+                .get(
+                    "/validator/attestation",
+                    processGetRequestImpl(this::produceAttestationResponse))
+                .post(
+                    "/validator/attestation", processPostRequestImpl(this::acceptAttestationSubmit))
+                .get("/node/fork", doJsonResponse(this::produceForkResponse)));
+  }
+
+  private String produceVersionResponse() {
+    if (version == null) {
+      Properties props = loadResources("rest-server.properties", this.getClass().getClassLoader());
+      final String version = props.getProperty("versionNumber");
+      this.version = String.format("Beacon Chain Java %s", version);
+    }
+    return version;
+  }
+
+  private Properties loadResources(final String name, final ClassLoader classLoader) {
+    try {
+      final Enumeration<URL> systemResources =
+          (classLoader == null ? ClassLoader.getSystemClassLoader() : classLoader)
+              .getResources(name);
+      Properties props = new Properties();
+      props.load(systemResources.nextElement().openStream());
+
+      return props;
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private String produceSyncingResponse() {
+    try {
+      return mapper.writeValueAsString(syncingResponse);
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private String produceForkResponse() {
+    try {
+      ForkResponse forkResponse =
+          new ForkResponse(
+              observableBeaconState.getLatestSlotState().getFork().getCurrentVersion().toString(),
```

Comment:
There could be a race condition here in a case when `observableBeaconState` is updated somewhere in between of these subsequent calls.

Reply:
I see such simple resolution for it: https://github.com/harmony-dev/beacon-chain-java/commit/74342ba7b0735bb58a6da3e9f7b9bb7b0e88e316
but I don't like it much, because it's running on one thread which is used for server answers and shouldn't be blocked. @Nashatyrev do you have any idea for such cases?

Reply:
In this concrete case (no blocking calls, no calls to external components) both solutions are fine:
- synchronize both methods (as you already did in `ForkController`)
- just assign `observableBeaconState` to a local variable